README.md
# DiscordServerBot

## A Discord server bot that can be run inside a Docker container

### Downloading

You can download the project in one of three ways: by cloning the project from
[GitHub](https://github.com/Ryan-M-Smith/DiscordServerBot), by pulling the latest Docker image from my
[Docker Hub](https://hub.docker.com/r/ryanmsmith144/discordserverbot), or by pulling an image from [GitHub
Packages](https://github.com/Ryan-M-Smith?tab=packages&repo_name=DiscordServerBot).

```sh
# Clone from GitHub
git clone https://github.com/Ryan-M-Smith/DiscordServerBot.git

# Pull from Docker Hub (pulls tag latest by default)
docker pull ryansmith144/discordserverbot

# To pull a specific tag (I don't suggest using latest):
docker pull ryanmsmith144/discordserverbot:<tagname>

# To pull from GitHub packages.
docker pull ghcr.io/ryan-m-smith/discordserverbot:<tagname>
```

### Prerequisites

**Python Version:** If you use the Docker container, by default the code is built using Python 3.6. I chose
to default it to this to support users with older Python versions. You can change this and use a newer version
simply by modifying the Dockerfile and uncommenting the version you want. Be sure to comment out the other one though,
unless you want to fetch more than one Python version.

**Packages:** To run the bot, you'll need to have all of the `pip`/PyPi packages installed. If you plan to run the
code with Docker, you'll also need to have  `docker` and `docker-compose` installed. No matter what method you choose,
all `pip` requirements will be automatically installed (by either `setup.py` or `docker-compose`).

### Supported Platforms

The code should work fine on Linux, macOS, and Windows computers.

### Running

To run a Discord bot, you need a Discord account and a registered Discord application. You'll also need access to a server
where you have member access or higher. There's a tutorial [here](https://discordpy.readthedocs.io/en/latest/discord.html)
that can help you with all the setup (i.e., getting a bot set up on your server). Technically, this isn't a bot. It's code
that runs using a bot on your server and allows it to do certain things.

Before you run the project, you need to put your bot's token and the name of your server into the `.env` file
in the project's root directory. When you do this, your `.env` should look like this:

```sh
DISCORD_TOKEN=<YOUR_TOKEN> # There shouldn't be quotes around this
DISCORD_SERVER="My Server's Name" # Quotes are okay here
```

If you're using the Docker image, just reproduce the `.env` file above in your home directory, or wherever is most
convenient for you.

If you plan to use any method involving Docker, consider these things:

* Make sure `docker` and `docker-compose` are installed on your system
* Enable the `docker` service with `sudo systemctl enable --now docker`
* If you want to be able to run Docker without sudo, run `sudo groupadd docker` then
  `sudo usermod -aG docker $USER`

Once you have everything downloaded and set up, you can choose to run the project by:

1. Pull and run an image of the Docker container from Docker Hub
    * Run `docker run -p 80:80 ryanmsmith144/discordserverbot:<TAGNAME> --env-file /path/to/env/file`
      * You can use whatever port and tag you choose
      * You can make a `.env` file on your system and put the necessary information in there,
        or you can use `--env` and pass in everything in the command itself. If you plan to
        make your own `.env` file, see the one in the root directory of this project for a
        template.

2. Building an image of the project's Docker container and running it

    * Clone the repo and `cd` into the directory
    * Run `docker-compose up` to build and run the project
      * The code runs forever, so use `ctrl+c` to exit
      * Once you close the container, use `docker run --env-file .env serverbot` to run
      * Should you ever need to rebuild, run `docker-compose up --build`

3. Using `setup.py install` and running a Python application

    * Run `python3 setup.py install`
      * Run this as root (`sudo`) to install the program in the `/usr/local/lib/.../site-packages`
      * Run this with the `--user` flag to install the program in `~/.local/lib/.../site-packages`
    * Run `serverbot` in your terminal

### Capabilities

The bot can perform actions within your server (like greeting users) as well as custom slash (/)
commands. To learn more about what the bot can do, see
[usage.md](https://github.com/Ryan-M-Smith/DiscordServerBot/blob/master/doc/usage.md).

### More Info

To learn more about the project, see
[software-info.md](https://github.com/Ryan-M-Smith/DiscordServerBot/blob/master/doc/software-info.md).


CHANGELOG.md
# DiscordServerBot Changelog

This changelog follows the basic format outlined on [keepachangelog.com](https://keepachangelog.com/en/1.0.0).

## v0.2.0 - Released 2021-07-07

* **Added**
  * Added a Rage role in addition to the Zen role
  * The bot has a status now
  * Added integration with GitHub packages

* **Changed**
  * The roles now have colors
  * Updated some documentation

## v0.1.1 - Released 2021-06-16

* **Added**
  * There are now import statements for `setup.py` and Docker

* **Changed**
  * Updated some documentation
    * Added info about a way to run the Docker container without pulling it
  * Updated the dockerignore

## v0.1.0 - Released 2021-06-13

* The initial release of DiscordServerBot


src/bot.py
#
# FILENAME: bot.py | ServerBot
# DESCRIPTION: Allow the bot to connect and interface with Discord
# CREATED: 2021-06-03 @ 8:08 PM
# COPYRIGHT: Copyright (c) 2021 by Ryan Smith <rysmith2113@gmail.com>
#

import os, sys
from pathlib import Path
from typing import NoReturn
from multiprocessing import Process, Event

import discord
from dotenv import load_dotenv

# NOTE: Import differently for Docker builds
#import commands
from . import commands

# Load the tokens from the environment
load_dotenv(Path(".env"))
TOKEN = os.getenv("DISCORD_TOKEN")
SERVER = os.getenv("DISCORD_SERVER")

intents = discord.Intents.default()
intents.members = True
client = discord.Client(intents=intents)

@client.event
async def on_ready() -> NoReturn:
	""" A "ready" message. """
	print(f"{client.user} is online")
	channel = client.get_channel(850104153278644305)
	await channel.send(f"{client.user} hopped online")

	# Show that the bot is ready
	await client.change_presence(
		status=discord.Status.idle,
		activity=discord.Game(name='Use /help for help')
	)

@client.event
async def on_member_join(member: discord.Member) -> NoReturn:
	""" Display a message when a new user joins the server """
	channel = client.get_channel(850104153278644305)
	await member.create_dm()
	await channel.send(f"What's poppin\', {member.name}? Welcome to Infinitely Sus!")

def main(event: Event) -> NoReturn:
	""" Run the server events. """

	event.wait()
	print("Server events active.")
	client.run(TOKEN)

def run() -> NoReturn:
	""" Simultaneously run the server events and the slash commands. """

	py_version = sys.version_info

	# Print some info to the terminal 
	print("DiscordServerBot - Copyright (c) 2021 by Ryan Smith")
	print(f"Running with Python {py_version.major}.{py_version.minor}.{py_version.micro}")
	print("-" * 51) # A divider (creates 51 dashes)

	event = Event()

	# Get both file's main functions
	bot, slash_commands = Process(target=main, args=tuple([event])), Process(target=commands.main, args=tuple([event]))

	# Start the processes
	bot.start()
	slash_commands.start()

	event.set() # Start the event

if __name__ == "__main__":
	sys.exit(run())


src/params.py
#
# FILENAME: params.py | ServerBot
# DESCRIPTION: Parameter lists (list of dictionaries) for commands
# CREATED: 2021-06-04 @ 12:23 AM
# COPYRIGHT: Copyright (c) 2021 by Ryan Smith <rysmith2113@gmail.com>
#

PING_OPTIONS = [
	{
		"name": "times",
		"description": "The number of times to ping",
		"required": False,
		"type": 4
	}
]

GREET_OPTIONS = [
	{
		"name": "member",
		"description": "The member to greet",
		"required": True,
		"type": 9
	}
]

EIGHTBALL_OPTIONS = [
	{
		"name": "question",
		"description": "A question to ask the magic 8-ball",
		"required": True,
		"type": 3
	}
]

DICEROLL_OPTIONS = [
	{
		"name": "count",
		"description": "The number of dice to roll",
		"required": False,
		"type": 4
	},

	{
		"name": "sides",
		"description": "The number of sides the die will have",
		"required": False,
		"type": 4
	}
]

RANDOM_OPTIONS = [
	{
		"name": "lower",
		"description": "The minimum number to show",
		"required": False,
		"type": 4
	},

	{
		"name": "upper",
		"description": "The maximum number to show",
		"required": False,
		"type": 4
	}
]

src/commands.py
#
# FILENAME: commands.py | ServerBot
# DESCRIPTION: Custom slash (/) commands
# CREATED: 2021-06-03 @ 10:10 PM
# COPYRIGHT: Copyright (c) 2021 by Ryan Smith <rysmith2113@gmail.com>
#

from multiprocessing import Event
import os, secrets, sys

from typing import List, NoReturn, Optional
from random import choice, randint, randrange
from datetime import datetime, date
from pathlib import Path

# NOTE: Import differently for Docker builds
#from params import * 
from .params import *

import discord

from discord import Member
from discord.ext import commands
from discord.utils import get
from discord_slash import SlashCommand, SlashContext, SlashCommandOptionType
from dotenv import load_dotenv

load_dotenv(Path(".env"))
TOKEN = os.getenv("DISCORD_TOKEN")
SERVER = os.getenv("DISCORD_SERVER")

intents = discord.Intents.all()
intents.members = True

client = commands.Bot(command_prefix="/", intents=intents)
slash = SlashCommand(client)

@client.command(
	name="ping", description="Ping the server.",
	guild_ids=[850104152679252006], options=PING_OPTIONS
)
async def ping(ctx: SlashCommand, times: int = 1) -> NoReturn:
	"""
		Ping the server.
		Usage: `/ping <times?>`
	"""
	for _ in range(times):
		await ctx.send(content=f"Query recieved ({round(client.latency * 1000)} ms)")

@client.command(
	name="info", description="Get info about the server.",
	guild_ids=[850104152679252006], options=None
)
async def info(ctx: SlashCommand) -> NoReturn:
	"""
		Get info about the server.
		Usage: `/info`
	"""
	time, curr_date = datetime.now(), date.today()
	slice = 0 if int(time.strftime("%I")) >= 10 else 1

	await ctx.send(content=
		f"Welcome to the {SERVER} server!\n" + \
		f"It is currently {curr_date.strftime('%B %d, %Y')} at {time.strftime('%I:%M %p')[slice:]}"
	)

@client.command(
	name="greet", description="Greet a user.",
	guild_ids=[850104152679252006], options=GREET_OPTIONS
)
async def greet(ctx: SlashCommand, member: Member) -> NoReturn:
	"""
		Greet a user.
		Usage: `/greet @<member: 9>`
	"""
	await ctx.send(f"Greetings, {member.name}! How's life?")

@client.command(
	name="8ball", description="Simulate a magic 8-ball.",
	guild_ids=[850104152679252006], options=EIGHTBALL_OPTIONS
)
async def eight_ball(ctx: SlashCommand, _: str) -> NoReturn: # The command input (question) is just for looks, and is ignored here
	"""
		Simulate a magic 8-ball
		Usage: `/8ball <question: 3>`
	"""

	#
	# The typical 8-ball responses, found at https://en.wikipedia.org/wiki/Magic_8-Ball#Possible_answers.
	# There are 20 possible answers, comprised of:
	#	* 10 affirmative
	#	* 5 non-commital
	#	* 5 negative
	#
	RESPONSES = [
		# Affirmative
		"It is Certain.", "It is decidedly so.", "Without a doubt.", "Yes definitely.",
		"You may rely on it.", "As I see it, yes.", "Most likely.", "Outlook good.",
		"Yes.", "Signs point to yes.",

		# Non-commital
		"Reply hazy, try again.", "Ask again later.", "Better not tell you now.",
		"Cannot predict now.", "Concentrate and ask again.",

		# Negative
		"Don't count on it.", "My reply is no.", "My sources say no.",
		"Outlook not so good.", "Very doubtful."
	]

	await ctx.send(f"Magic 8-ball: {choice(RESPONSES)}")

@client.command(
	name="roll", description="Roll a user-specified amount of dice of a user-specified side count.",
	guild_ids=[850104152679252006], options=DICEROLL_OPTIONS
)
async def diceroll(ctx: SlashCommand, count: int = 1, sides: int = 6) -> NoReturn:
	"""
		Roll a die (or dice). The user can specify the amount of dice to roll as well
		as how many sides each die should have. This command allows the user to roll a
		die with a side count of 4, 6, 8, 10, 12, or 20.
		Usage: `/roll <count?: 4 = 1> <sides?: 4 = 6>
	"""

	def display(_list: List[int]) -> str:
		""" Customize how a list is displayed. """
		return ', '.join(list(map(lambda n: str(n), _list)))

	DIE_SIDE_COUNTS = [4, 6, 8, 10, 12, 20]

	# Rolling 0 dice is useless, and rolling negative dice is impossible 
	if count <= 0:
		await ctx.send(f"Dice roll: not possible to roll {count} dice")
		return
	
	# Only certain side counts are allowed to be rolled
	if sides not in DIE_SIDE_COUNTS:
		await ctx.send(f"Dice roll: rolling a {sides}-sided die is not allowed")
		return
	
	await ctx.send(f"Rolling {count} {sides}-sided {'dice' if count > 1 else 'die'}") # Maintain grammatical correctness :)

	result = list()

	# Calculate the results
	for _ in range(count):
		result.append(randrange(1, sides))

	await ctx.send(f"Result: {display(result)}") # Display the answer

@client.command(
	name="rage", description="Expell rage in the server.",
	guild_ids=[850104152679252006], options=None,
	pass_context=True
)
async def rage(ctx: SlashCommand) -> NoReturn:
	"""
		Rage on the server.
		Usage: `/rage`
	"""

	member = ctx.message.author # The command sender
	role_rage = discord.utils.get(ctx.message.guild.roles, name="Rage")
	role_zen = discord.utils.get(ctx.message.guild.roles, name="Zen")

	# Remove the Zen role from the user, if it exists and the user has it
	if get(ctx.guild.roles, name="Zen") and (role_zen in member.roles):
		await member.remove_roles(role_zen)
	
	# Create the Rage role if one doesn't already exist
	if not get(ctx.guild.roles, name="Rage"):
		await ctx.guild.create_role(name="Rage")
		role_rage = discord.utils.get(ctx.message.guild.roles, name="Rage") # Get the role if it didn't exist before
	
	await role_rage.edit(name="Rage", color=discord.Color(int(0xff2a00))) # Change the role color
	await ctx.message.author.add_roles(role_rage) # Give the user the Rage role

	await ctx.send(f"{member} is raging!")

	for _ in range(randint(10, 15)):
		await ctx.send(secrets.token_hex(nbytes=16)) # Simulate keymashing

@client.command(
	name="calm", description="Be calm in the server.",
	guild_ids=[850104152679252006], options=None,
	pass_context=True
)
async def calm(ctx: SlashCommand) -> NoReturn:
	"""
		Enter a calm (zen) state on the server
		Usage: `/calm`
	"""

	member = ctx.message.author # The command sender
	role_rage = discord.utils.get(ctx.message.guild.roles, name="Rage")
	role_zen = discord.utils.get(ctx.message.guild.roles, name="Zen")

	# Remove the Rage role from the user, if it exists and the user has it
	if get(ctx.guild.roles, name="Rage") and (role_rage in member.roles):
		await member.remove_roles(role_rage)

	# Create the Zen role if one doesn't already exist
	if not get(ctx.guild.roles, name="Zen"):
		await ctx.guild.create_role(name="Zen")
		role_zen = discord.utils.get(ctx.message.guild.roles, name="Zen") # Get the role if it didn't exist before
	
	await role_zen.edit(name="Zen", color=discord.Color(int(0x009282))) # Change the role color
	await member.add_roles(role_zen) # Give that user the Zen role
	
	await ctx.send(f"{member} is being calm. Do not disturb their zen state.")

@client.command(
	name="random", description="Display a random number.",
	guild_ids=[850104152679252006], options=RANDOM_OPTIONS
)
async def random(ctx: SlashCommand, lower: int = -sys.maxsize, upper: int = sys.maxsize):
	"""
		Print a random number in the range [lower, upper]
		Usage: `/random <lower?: 4 = -sys.maxsize> <upper?: 4 = sys.maxsize>
	"""

	await ctx.send(f"Your random number is: {randint(lower, upper)}")

def main(event: Event) -> NoReturn:
	""" Run the slash commands. """

	event.wait()
	print("Slash commands loaded.")
	client.run(TOKEN)
